stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -673,38 +690,16 @@ function gemm_wrapper!(C::StridedVecOrMat{T}, tA::AbstractChar, tB::AbstractChar
         stride(C, 2) >= size(C, 1))
         return BLAS.gemm!(tA, tB, alpha, A, B, beta, C)
     end
-    generic_matmatmul!(C, tA, tB, A, B, _add)
+    # Not using @stable_muladdmul here deliberately to create an inferrence
+    # barrier in case α, β are not compile-time constants. This avoids compiling
+    # four versions of generic_matmatmul!() in those cases.
```

Comment:
As already questioned in the comments, I think these changes shouldn't be here (i.e. all the changes that tries to give special treatment to the general method, for both syrk, herk, and gemm wrappers).

1. i see no reason why one would want to preserve the type instability it definitely introduces; It basically just solves the problem for 2x2 and 3x3 matrices (for which one should probably use a static array instead). Code is much more difficult to understand since it makes this strange exception to the pattern otherwise used for all other types.
2. the inline and especially noinline are just hints (that the compiler can and often will decide what to do regardless), though here they are used as if they introduce some strong guarantees.
3. all of these changes will have to be reverted back once a nicer solution with staticbool can be used.

Reply:
As for the first concern, runtime dispatch is not a bad ideal when the inner kernal is big. Although we can make it stable, but the complie tex would be high.

I agree that 2x2 and 3x3 are not that important. And in most cases, these repeated branches would never be hit.
But the current design was caused by their performance regression. So I'm not sure is it ok to re-introduce the regression.

As for the 3rd concern, there's no time line for staticbool and related compiler improvement. #46471 might help here, but I don't think it's a block for temporary improvement.

Reply:
The only purpose of this PR is to replace the runtime dispatch with a static if-else block at the cost of increased compilation times. Except it excludes the one scenario which everyone cares about, gemm for matrices larger than 3x3

I'm not suggesting dropping the small matrix optimizations, but this PR would literally keep the instability in the case that actually started the #46865, only making it slightly worse since it needs to compile more for the small matrix variants. And it moves the type instability deeper, so it seems less likely that optimizations would take the new instability away.

I think const prop is a red herring; it would of course not do anything in the common case where alpha and beta aren't constant.

Reply:
Except that it does *not* exclude the case of gemm for matrices larger than
3x3. The BLAS case is explicitly included: it no longer involves MulAddMul at
all in this PR. This PR fixes the allocations seen in #46865 - it even includes
an explicit test for this!

Also, compile times are only worse in cases that previously involved
runtime dispatch, since const-prop should still eliminate the branches
(assuming the `@inline` hint is successful -- see below).

`generic_matmatmul` is the only case that now involves runtime dispatch
(likely always) but it is also *not* the hot path everyone cares about, as
far as I am aware. When I tested it with dense matrices, it did allocations
all by itself and the runtime dispatch overhead was negligible.

Regarding inlining hints: They are not required to be respected for this
approach to work. The `@inline` is the more important one, as it makes it
more likely for const-prop to still happen in the small-matrix cases and
eliminate the new branches (keeping compile time down). The `@noinline`
doesn't really serve a useful purpose beyond showing the reader why
the `gemm_wrapper` functions are split the way they are and could just
be removed.

Reply:
If there are no replies, can I assume this is resolved?

Reply:
@Micket Reading [this comment](https://github.com/JuliaLang/julia/pull/44538#issuecomment-1382174693) made me realize that the changes in this PR, plus use of Static.jl, pretty much give you the interface you have been suggesting. If you use Static.jl you can already supply `alpha` and `beta` as static bools or ints, which will lead to compile-time selection of the correct methods as the branches added in this PR will be evaluated at compile time. The only sacrifice is the `generic_matmatmul` inference barrier, which doesn't appear to hurt performance noticeably.

Reply:
Yes but introducing the static half of the problem was never really any issue (whether you used Union{Nothing,Number}, Val{x}, static(x), or just introduced your own custom type (which I played around with, only takes a few lines of code)). It was to stop doing all the extra stuff for the dynamic case which brought the real benefits I.M.H.O.

A `beta=static(0)` would allow the compiler to maybe optimize away a branch, but for the general case, even if we know we have `beta=rand()+1` (something strictly > 0), the extra compilations for all those possible runtime branches will still be there (probably, I don't think the optimizer would see through that).
But in both of those if your matrices are small enough to notice a runtime cost from this you should surely be using StaticArrays anyway. 

I just think the interface was wrong from the start; to select algorithm based on value when it could/should have been based *only* on type, which is necessarily a breaking change. 
The main benefits i see from that is code simplification, no special macros, no extra branches and special rules. You simply get what you call for, and only ever pay for that.

(I'm also not sure static(1)/static(0) is better at Nothing here; since what one is really selecting a different algorithm in a way; not just some optimization as it's not really equal to multiplication by 0 anymore (in terms of IEEE-754 math))


Note: I've gotten beyond busy recently, so I'm unlikely to write more in this issue. I don't have any authority here anyway and anyone should feel free to dismiss my comments.